File: prospector/facets.py
"""The discovery facet vocabulary — the Python end of one closed contract.

The same six vocabularies exist in C# at
``store_platform/src/Store.Catalog/Domain/PackFacets.cs`` and in TypeScript at
``store_platform/src/Store.Web/src/lib/facets.ts``. Three copies is a deliberate cost: the
engine, the API, and the browser are three deploy units, and a shared runtime dependency
between them would be a worse coupling than three lists that a test can compare.

Two rules this module exists to enforce:

1. **Absent means absent.** A value the engine cannot justify is ``None``, never a default
   and never a guess. ``normalize`` drops anything outside the vocabulary rather than
   coercing it to the nearest member — a coerced facet is a claim nobody made, and the
   buyer would filter on it believing it came from the dossier.

2. **The legacy ``effort_tag`` is not a source for ``effort``.** ``low | medium | high``
   was never defined to mean "how much of delivery is machine-doable"; mapping ``high`` to
   ``hands_on`` is a guess wearing the costume of a migration. See spec 2.3.
"""

from __future__ import annotations

from typing import Any, Dict, Iterable, List, Optional

#: What the buyer already has. Multi-valued (0-3) — the primary router input.
ADVANTAGE = ("code", "nocode", "sales", "ops", "audience")

#: Who signs the cheque.
PAYER = ("b2b", "b2c", "b2g")

#: How much of delivery is machine-doable.
EFFORT = ("automatable", "part_automatable", "hands_on")

#: Hours needed to RUN it — deliberately separate from EFFORT.
COMMITMENT = ("evenings", "part_time", "full_time")

#: How it makes money. Mirrors config.yaml:595-603 (``generation.structural_forms``).
MECHANISM = (
    "productized_service",
    "vertical_tool",
    "transaction_broker",
    "risk_financing",
    "physical_ops",
    "audience_media",
    "picks_and_shovels",
    "data_intelligence",
)

#: Display and exclusion only ("anything but vets") — never the primary filter.
SECTOR = (
    "licensing_admin",
    "employment_pay",
    "housing_rental",
    "care_benefits",
    "trades_construction",
    "pets_animals",
    "creative_rights",
    "property_probate",
    "energy_planning",
    "retail_inventory",
    "professional_services",
    "other",
)

#: Single-valued facets and their vocabularies, in the order they appear on the wire.
SINGLE_VALUED = {
    "sector": SECTOR,
    "payer": PAYER,
    "effort": EFFORT,
    "commitment": COMMITMENT,
    "mechanism": MECHANISM,
}

MAX_ADVANTAGES = 3


def clean_one(value: Any, allowed: Iterable[str]) -> Optional[str]:
    """Return ``value`` if it is in ``allowed``, else ``None``.

    Case- and whitespace-tolerant on the way in, because an operator returning
    ``" B2B "`` meant ``b2b`` and there is no ambiguity to resolve. Everything else is
    dropped: ``"business"`` is not ``b2b``, it is an unrecognised answer, and guessing
    which member it meant is exactly the inference this contract forbids.
    """
    if value is None:
        return None
    text = str(value).strip().lower()
    return text if text in tuple(allowed) else None


def clean_advantages(values: Any) -> List[str]:
    """Normalise the multi-valued advantage list, de-duplicated and capped.

    Unknown members are dropped individually rather than failing the list: unlike the
    publish API — which rejects the whole request so a partial write cannot happen — this
    runs at generation time, where keeping the two advantages the model justified is
    strictly better than discarding all three because it invented a fourth.
    """
    if not isinstance(values, (list, tuple, set)):
        return []
    out: List[str] = []
    for v in values:
        cleaned = clean_one(v, ADVANTAGE)
        if cleaned and cleaned not in out:
            out.append(cleaned)
    return out[:MAX_ADVANTAGES]


def normalize(raw: Any) -> Dict[str, Any]:
    """Coerce a model-supplied ``facets`` object into the contract.

    Always returns all six keys. Single-valued facets are ``None`` when absent or
    unrecognised; ``advantages`` is ``[]``. Never raises: a malformed facets block must
    cost the listing its tags, not the whole publish.
    """
    if not isinstance(raw, dict):
        raw = {}

    facets: Dict[str, Any] = {
        name: clean_one(raw.get(name), allowed)
        for name, allowed in SINGLE_VALUED.items()
    }
    facets["advantages"] = clean_advantages(raw.get("advantages"))
    return facets


def to_wire(facets: Dict[str, Any]) -> Dict[str, Any]:
    """Map the snake_case engine facets onto the camelCase publish payload.

    Empty values are dropped so the Store API's only-overwrite-when-sent rule applies: a
    facet-light republish must never silently untag a pack the backfill tagged.
    """
    # The Store API binds JSON property names case-insensitively, so these lower-case
    # names land on PublishRequest.Sector/Payer/… Spelled out rather than derived, so a
    # rename on either side breaks visibly here instead of silently dropping a facet.
    wire: Dict[str, Any] = {}
    for name in ("sector", "payer", "effort", "commitment", "mechanism"):
        value = facets.get(name)
        if value:
            wire[name] = value
    advantages = facets.get("advantages") or []
    if advantages:
        wire["advantages"] = list(advantages)
    return wire

File: store_platform/src/Store.Web/src/lib/facets.ts
/**
 * The discovery facet vocabulary — the TypeScript end of one closed contract.
 *
 * The same six vocabularies exist in Python at `prospector/facets.py` and in C# at
 * `store_platform/src/Store.Catalog/Domain/PackFacets.cs`. Three copies is a deliberate cost:
 * the engine, the API and the browser are three deploy units, and a shared runtime dependency
 * between them would be a worse coupling than three lists a test can compare. The test that
 * makes the cost safe is `src/lib/__tests__/facets.test.ts`, which reads `PackFacets.cs` off
 * disk and asserts value-for-value equality — so a facet added in one language and forgotten
 * in another fails `npm test` rather than silently disappearing from the filter bar.
 *
 * Two rules this module exists to enforce:
 *
 * 1. **No facet is ever inferred from pack text.** The storefront guessed sector with a regex
 *    over title + one-liner and told buyers a metal-fabrication quoting engine was a gardening
 *    business (`src/lib/category.ts`, deleted by this story). On a brand whose whole position
 *    is "every claim sourced", a filter that lies is worse than no filter.
 * 2. **Absent means absent.** An untagged pack renders no chip and appears only under "All".
 *    There is no default value anywhere in this file, and `label()` returns null for an
 *    unknown code rather than prettifying it — a rendered guess is a claim nobody made.
 */

export const ADVANTAGE = ['code', 'nocode', 'sales', 'ops', 'audience'] as const;
export const PAYER = ['b2b', 'b2c', 'b2g'] as const;
export const EFFORT = ['automatable', 'part_automatable', 'hands_on'] as const;
export const COMMITMENT = ['evenings', 'part_time', 'full_time'] as const;
export const MECHANISM = [
  'productized_service',
  'vertical_tool',
  'transaction_broker',
  'risk_financing',
  'physical_ops',
  'audience_media',
  'picks_and_shovels',
  'data_intelligence',
] as const;
export const SECTOR = [
  'licensing_admin',
  'employment_pay',
  'housing_rental',
  'care_benefits',
  'trades_construction',
  'pets_animals',
  'creative_rights',
  'property_probate',
  'energy_planning',
  'retail_inventory',
  'professional_services',
  'other',
] as const;

export type Advantage = (typeof ADVANTAGE)[number];
export type Payer = (typeof PAYER)[number];
export type Effort = (typeof EFFORT)[number];
export type Commitment = (typeof COMMITMENT)[number];
export type Mechanism = (typeof MECHANISM)[number];
export type Sector = (typeof SECTOR)[number];

/** A pack carries at most three advantages (`prospector/facets.py` MAX_ADVANTAGES). */
export const MAX_ADVANTAGES = 3;

/** Every facet except `advantage` holds one value or none. */
export const SINGLE_VALUED_KINDS = ['sector', 'payer', 'effort', 'commitment', 'mechanism'] as const;
export type SingleValuedKind = (typeof SINGLE_VALUED_KINDS)[number];
export type FacetKind = SingleValuedKind | 'advantage';

export const VOCABULARY: Record<FacetKind, readonly string[]> = {
  advantage: ADVANTAGE,
  sector: SECTOR,
  payer: PAYER,
  effort: EFFORT,
  commitment: COMMITMENT,
  mechanism: MECHANISM,
};

/** True when `value` is a member of that facet's closed vocabulary. Null/undefined are false. */
export function isFacetValue(kind: FacetKind, value: string | null | undefined): boolean {
  if (!value) return false;
  return VOCABULARY[kind].includes(value);
}

/**
 * Display copy. Chips must read as English — the old card rendered `{effortTag} effort`, giving
 * buyers "Highly automatable effort" (spec Part 10). Every string here is buyer-facing and this
 * file is its only home, so a wording change happens in one place rather than five components.
 */
const LABELS: Record<FacetKind, Record<string, string>> = {
  advantage: {
    code: 'Suits builders',
    nocode: 'No code needed',
    sales: 'Suits sellers',
    ops: 'Suits operators',
    audience: 'Suits an audience',
  },
  payer: {
    b2b: 'Sells to businesses',
    b2c: 'Sells to consumers',
    b2g: 'Sells to public bodies',
  },
  effort: {
    automatable: 'Mostly automated',
    part_automatable: 'Part automated',
    hands_on: 'Hands-on service',
  },
  commitment: {
    evenings: 'Evenings-friendly',
    part_time: 'Part-time hours',
    full_time: 'Full-time',
  },
  mechanism: {
    productized_service: 'Productised service',
    vertical_tool: 'Vertical tool',
    transaction_broker: 'Transaction broker',
    risk_financing: 'Risk and financing',
    physical_ops: 'Physical operations',
    audience_media: 'Audience and media',
    picks_and_shovels: 'Picks and shovels',
    data_intelligence: 'Data intelligence',
  },
  /**
   * Sector display copy. The old set read as a government directory ("Licensing and admin",
   * "Employment and pay", "Other") — accurate, but it described the paperwork rather than the
   * opportunity, and a buyer scanning a shelf of £49 packs has no reason to click "Other".
   *
   * These are LABELS ONLY. The codes below are the closed cross-language contract shared with
   * `prospector/facets.py` and `PackFacets.cs`; nothing here renames, merges or splits a code.
   * That matters because the obvious "improvement" is to merge — e.g. one "worker justice" tag
   * over both `employment_pay` and `care_benefits`. That would silently make the three
   * vocabularies disagree and break the filter contract, so it is not done here: two codes stay
   * two labels.
   *
   * The bar every string had to clear: punchier tone, but still only describing WHAT THE SECTOR
   * IS. No label may assert an outcome the dossier has not proven. "Niche cash cows" for `other`
   * and "Lucrative" anything are rejected on exactly that basis — profitability is a claim, and
   * on a storefront whose position is "every claim sourced", a category cannot be the one place
   * we assert something with no source behind it.
   */
  sector: {
    licensing_admin: 'Red-tape and licensing',
    employment_pay: 'Pay and worker rights',
    housing_rental: 'Housing and tenancy',
    care_benefits: 'Care and benefits claims',
    trades_construction: 'Trades and site work',
    pets_animals: 'The pet economy',
    creative_rights: 'Creator rights',
    property_probate: 'Property and probate',
    energy_planning: 'Energy and planning',
    retail_inventory: 'Retail and stock',
    professional_services: 'Professional services',
    other: 'Specialist niches',
  },
};

/**
 * Compact copy for dense rows — the command palette shows four facts plus a price on one line
 * (spec :274), where "Sells to consumers" pushes the price off a phone. Only defined where it
 * differs from the full label; `shortLabel` falls back to `label` otherwise.
 */
const SHORT_LABELS: Partial<Record<FacetKind, Record<string, string>>> = {
  payer: { b2b: 'B2B', b2c: 'B2C', b2g: 'B2G' },
  advantage: {
    code: 'Builders',
    nocode: 'No code',
    sales: 'Sellers',
    ops: 'Operators',
    audience: 'Audience',
  },
};

/**
 * Buyer-facing text for one facet value, or null when the value is absent or outside the
 * vocabulary. Null is deliberate: an unrecognised code means the API is ahead of this deploy,
 * and rendering the raw token ("part_automatable") or a prettified guess would both be worse
 * than rendering nothing.
 */
export function label(kind: FacetKind, value: string | null | undefined): string | null {
  if (!value) return null;
  return LABELS[kind][value] ?? null;
}

/** As `label`, but the compact form where one exists. Same null contract. */
export function shortLabel(kind: FacetKind, value: string | null | undefined): string | null {
  if (!value) return null;
  return SHORT_LABELS[kind]?.[value] ?? label(kind, value);
}

/**
 * Heading copy for a facet's group in the filter bar. Each one names the question the buyer is
 * answering about themselves, not the database column — "Sector" told a buyer nothing about what
 * clicking would do, and "What you already have" left "have of what?" unanswered.
 *
 * These are HEADINGS ONLY: they sit above a group of chips and are never dropped into a sentence.
 * Use `KIND_NOUN` for that — see below for why the distinction is not cosmetic.
 *
 * Revised 2026-08-01 toward a tighter register, per string rather than wholesale. The sidebar is
 * `lg:grid-cols-[15rem_1fr]` (`pages/index.tsx:428`) and these render uppercase with
 * `tracking-widest`, so a four-word heading wraps to two lines and the group reads as heavier than
 * the chips underneath it. Three headings were replaced with the shorter conventional term; three
 * were kept, and the reason each was kept is that the shorter alternative was jargon for the same
 * idea rather than a clearer name for it:
 *
 * - `effort` keeps "How much is automated" over "Tech Enablement", which does not tell a buyer
 *   what the options underneath it will say.
 * - `mechanism` keeps "How it makes money" over "Revenue Model" — identical length, and one of
 *   them is readable by someone who has never worked in a startup.
 * - `advantage` takes "Skills you have" rather than "Founder Skillset": shorter than the original,
 *   but without the word "Founder", because the `evenings` commitment lane exists precisely for
 *   buyers who do not yet call themselves that.
 */
export const KIND_LABEL: Record<FacetKind, string> = {
  advantage: 'Skills you have',
  payer: 'Target market',
  effort: 'How much is automated',
  commitment: 'Time commitment',
  mechanism: 'How it makes money',
  sector: 'Industry',
};

/**
 * The same six facets as a noun phrase that survives being embedded in a sentence.
 *
 * This exists because `KIND_LABEL` was being used in two sentence slots it cannot fit, and both
 * rendered broken English on a live page: `pages/index.tsx` builds the near-miss relax button as
 * `Show any ${...}` — "Show any what you already have", "Show any time it needs" — and
 * `components/discovery/EmptyState.tsx` builds "Not tagged for ${...}". A heading is a fragment
 * chosen to read well ABOVE a control; a noun is chosen to read well INSIDE a clause. One string
 * cannot be both, and trying made the copy worse every time a heading improved.
 */
export const KIND_NOUN: Record<FacetKind, string> = {
  advantage: 'skill set',
  payer: 'customer type',
  effort: 'automation level',
  commitment: 'time commitment',
  mechanism: 'revenue model',
  sector: 'sector',
};
